src/pure_aworset.erl

```diff
@@ -0,0 +1,257 @@
+            case orddict:find(Key, VV2) of
+                {ok, Value2} ->
+                    Acc andalso Value1 =< Value2;
+                error ->
+                    Acc andalso false
+            end
+        end,
+        true,
+        VV1
+     ) andalso not (VV1 == VV2).
+
+%% @doc Check if 2 POLogs are equal.
+-spec equal_polog(pure_type:polog(), pure_type:polog()) -> boolean().
+equal_polog(POLog1, POLog2) ->
+    orddict:size(POLog1) == orddict:size(POLog2) andalso
+    orddict:fold(
+        fun(Key, Value1, Acc) ->
+            case orddict:find(Key, POLog2) of
+                {ok, Value2} ->
+                    Acc andalso Value1 == Value2;
+                error ->
+                    Acc andalso false
+            end
+        end,
+        true,
+        POLog1
+     ).
+
+%% @doc Check redundancy `pure_aworset()'
+%% Called in remove_redundant().
+-spec redundant({pure_type:id(), pure_aworset_op()}, {pure_type:id(), pure_aworset_op()}) ->
+    integer().
+redundant({VV1, {add, Elem1}}, {VV2, {_X, Elem2}}) ->
+    case Elem1 == Elem2 andalso happened_before(VV1, VV2) of
+        true ->
+            1;
+        false ->
+            0
+    end.
+
+%% @doc Removes redundant operations from POLog of `pure_aworset()'
+%% Called upon updating (add, rmv) the `pure_aworset()'
+-spec remove_redundant_POLog({pure_type:id(), pure_aworset_op()}, pure_aworset()) -> pure_aworset().
+remove_redundant_POLog({VV1, Op}, {?TYPE, {POLog0, ORSet}}) ->
+    POLog1 = orddict:fold(
+        fun(Key, Value, Acc) ->
+            case redundant({Key, Value}, {VV1, Op}) of
+                0 ->
+                    orddict:store(Key, Value, Acc);
+                1 ->
+                    Acc
+            end
+        end,
+        orddict:new(),
+        POLog0
+    ),
+    {?TYPE, {POLog1, ORSet}}.
+
+%% @doc Removes redundant operations from POLog of `pure_aworset()'
+%% Called upon updating (add, rmv) the `pure_aworset()'
+-spec remove_redundant_Crystal({pure_type:id(), pure_aworset_op()}, pure_aworset()) -> {boolean(), pure_aworset()}.
+remove_redundant_Crystal({_VV1, {_X, Elem}}, {?TYPE, {POLog0, AWORSet}}) ->
+    case ordsets:is_element(Elem, AWORSet) of
+        true ->
+            {true, {?TYPE, {POLog0, ordsets:del_element(Elem, AWORSet)}}};
+        false ->
+            {false, {?TYPE, {POLog0, AWORSet}}}
+    end.
+
+
+%% @doc Removes redundant operations from POLog of `pure_aworset()'
+%% Called upon updating (add, rmv) the `pure_aworset()'
+-spec remove_redundant({pure_type:id(), pure_aworset_op()}, pure_aworset()) -> pure_aworset().
+remove_redundant({VV1, Op}, {?TYPE, {POLog, ORSet}}) ->
+    {Crystal_Changed, {?TYPE, {POLog0, PureAWORSet0}}} = remove_redundant_Crystal({VV1, Op}, {?TYPE, {POLog, ORSet}}),
+    case Crystal_Changed of
+        true ->
+            {?TYPE, {POLog0, PureAWORSet0}};
+        false ->
+            {?TYPE, {POLog1, PureAWORSet1}} = remove_redundant_POLog({VV1, Op}, {?TYPE, {POLog, ORSet}}),
```

Comment:
Why is this match needed?  Can't you just return the result of the function implicitly?